File: sthlmtraveling/src/main/java/com/markupartist/sthlmtraveling/ui/view/TicketDialogFragment.java
/*
 * Copyright (C) 2009-2016 Johan Nilsson <http://markupartist.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.markupartist.sthlmtraveling.ui.view;

import android.content.Context;
import android.net.Uri;
import android.os.Bundle;
import androidx.annotation.IntDef;
import androidx.annotation.Nullable;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.markupartist.sthlmtraveling.R;
import com.markupartist.sthlmtraveling.data.models.Fare;
import com.markupartist.sthlmtraveling.data.models.FareAttribute;
import com.markupartist.sthlmtraveling.utils.Analytics;
import com.markupartist.sthlmtraveling.utils.IntentUtil;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.ArrayList;
import java.util.List;


public class TicketDialogFragment extends BottomSheetDialogFragment {

    static final String ARG_FARE = "zones";

    public static TicketDialogFragment create(Fare fare) {
        Bundle b = new Bundle();
        b.putParcelable(ARG_FARE, fare);
        TicketDialogFragment fragment = new TicketDialogFragment();
        fragment.setArguments(b);
        return fragment;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.dialog_ticket, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        Bundle args = getArguments();
        Fare fare = args.getParcelable(ARG_FARE);

        RecyclerView recyclerView = (RecyclerView) view.findViewById(R.id.recycler_view);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));

        final TicketAdapter adapter = new TicketAdapter(getContext());
        for (FareAttribute fareAttribute : fare.getAttributes()) {
            CharSequence description = getText(fareAttribute.isReduced() ?
                    R.string.sms_ticket_price_reduced :
                    R.string.sms_ticket_price_full);
            adapter.addTicket(new TicketPrice(description,
                    fare.getZones(),
                    fareAttribute.getText(),
                    fareAttribute.isReduced(),
                    fareAttribute.getAction()));
        }

        recyclerView.setAdapter(adapter);
        ItemClickSupport.addTo(recyclerView).setOnItemClickListener(new ItemClickSupport.OnItemClickListener() {
            @Override
            public void onItemClicked(RecyclerView recyclerView, int position, View v) {
                TicketItem item = adapter.getItem(position);
                switch (item.viewType) {
                    case TicketItem.PRICE_ITEM:
                        TicketPrice ticketPrice = (TicketPrice) item.object;
                        Uri uri = Uri.parse(ticketPrice.getAction());
                        String action = uri.getQueryParameter("action");
                        if ("sms".equals(action)) {
                            sendSms(getContext(), ticketPrice.isReduced,
                                    uri.getQueryParameter("number"),
                                    uri.getQueryParameter("message"));
                        }
                        break;
                    case TicketItem.HEADER_ITEM:
                        // No-op
                        break;
                }
            }
        });
    }

    /**
     * Invokes the Messaging application.
     *
     * @param reducedPrice True if the price is reduced, false otherwise.
     * @param number       The number
     * @param message      The message
     */
    private static void sendSms(final Context context, final boolean reducedPrice,
                                final String number, final String message) {
        String price = reducedPrice ? "R" : "H"; // Only kept for analytic purposes
        Analytics.getInstance(context).event("Ticket", "Buy SMS Ticket", price);
        IntentUtil.smsIntent(context, number, message);
    }

    static class TicketPrice {
        public String zones;
        public CharSequence description;
        public boolean isReduced;
        public String priceText;
        public String action;

        public TicketPrice(CharSequence description,
                           String zones,
                           String priceText,
                           boolean isReduced,
                           String action) {
            this.description = description;
            this.zones = zones;
            this.priceText = priceText;
            this.isReduced = isReduced;
            this.action = action;
        }

        public CharSequence getPrice() {
            return priceText;
        }

        public String getAction() {
            return action;
        }
    }

    static class TicketAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

        private final List<TicketItem> mTicketItems = new ArrayList<>();
        private final LayoutInflater mLayoutInflater;
        private final Context mContext;

        public TicketAdapter(Context context) {
            this.mContext = context;
            this.mLayoutInflater = LayoutInflater.from(context);
        }

        @Override
        public int getItemViewType(int position) {
            return mTicketItems.get(position).viewType;
        }

        @Override
        public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            switch (viewType) {
                case TicketItem.PRICE_ITEM:
                    return PriceViewHolder.create(mLayoutInflater, parent);
            }
            return null;
        }

        @Override
        public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
            switch (getItemViewType(position)) {
                case TicketItem.PRICE_ITEM:
                    PriceViewHolder.bindTo((PriceViewHolder) holder,
                            (TicketPrice) mTicketItems.get(position).object);
                    break;
            }
        }

        @Override
        public int getItemCount() {
            return mTicketItems.size();
        }

        public void addTicket(TicketPrice ticket) {
            mTicketItems.add(new TicketItem<>(ticket, TicketItem.PRICE_ITEM));
        }

        public TicketItem getItem(int position) {
            return mTicketItems.get(position);
        }
    }

    public static class TicketItem<T> {

        @Retention(RetentionPolicy.SOURCE)
        @IntDef({HEADER_ITEM, PRICE_ITEM})
        public @interface ViewType {
        }

        public static final int HEADER_ITEM = 0;
        public static final int PRICE_ITEM = 1;

        public T object;
        @ViewType
        public int viewType;

        public TicketItem(T object, int viewType) {
            this.object = object;
            this.viewType = viewType;
        }
    }

    private static class PriceViewHolder extends RecyclerView.ViewHolder {
        private final TextView descriptionTextView;
        private final TextView priceTextView;
        private final TextView zonesTextView;

        public PriceViewHolder(View itemView) {
            super(itemView);
            descriptionTextView = (TextView) itemView.findViewById(R.id.ticket_description);
            priceTextView = (TextView) itemView.findViewById(R.id.ticket_price);
            zonesTextView = (TextView) itemView.findViewById(R.id.ticket_zones);
        }

        public static RecyclerView.ViewHolder create(LayoutInflater inflater, ViewGroup parent) {
            return new PriceViewHolder(inflater.inflate(R.layout.item_ticket_price, parent, false));
        }

        public static void bindTo(final PriceViewHolder holder, TicketPrice ticket) {
            holder.descriptionTextView.setText(ticket.description);
            holder.priceTextView.setText(ticket.getPrice());
            holder.zonesTextView.setText(ticket.zones);
        }
    }
}

File: sthlmtraveling/src/main/java/com/markupartist/sthlmtraveling/ui/view/TicketDialogFragment.kt
/*
 * Copyright (C) 2009-2016 Johan Nilsson <http://markupartist.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.markupartist.sthlmtraveling.ui.view

import android.content.Context
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.annotation.IntDef
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import com.markupartist.sthlmtraveling.R
import com.markupartist.sthlmtraveling.data.models.Fare
import com.markupartist.sthlmtraveling.utils.Analytics.Companion.getInstance
import com.markupartist.sthlmtraveling.utils.IntentUtil.smsIntent

class TicketDialogFragment : BottomSheetDialogFragment() {
    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.dialog_ticket, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        val args = getArguments()
        val fare = args!!.getParcelable<Fare?>(ARG_FARE)

        val recyclerView = view.findViewById<View?>(R.id.recycler_view) as RecyclerView
        recyclerView.setLayoutManager(LinearLayoutManager(getContext()))

        val adapter = TicketAdapter(getContext())
        for (fareAttribute in fare!!.attributes!!) {
            val description =
                getText(if (fareAttribute!!.isReduced) R.string.sms_ticket_price_reduced else R.string.sms_ticket_price_full)
            adapter.addTicket(
                TicketPrice(
                    description,
                    fare.zones,
                    fareAttribute.text,
                    fareAttribute.isReduced,
                    fareAttribute.action
                )
            )
        }

        recyclerView.setAdapter(adapter)
        ItemClickSupport.Companion.addTo(recyclerView)
            .setOnItemClickListener(object : ItemClickSupport.OnItemClickListener {
                override fun onItemClicked(recyclerView: RecyclerView?, position: Int, v: View?) {
                    val item = adapter.getItem(position)
                    when (item.viewType) {
                        TicketItem.Companion.PRICE_ITEM -> {
                            val ticketPrice = item.`object` as TicketPrice
                            val uri = Uri.parse(ticketPrice.action)
                            val action = uri.getQueryParameter("action")
                            if ("sms" == action) {
                                Companion.sendSms(
                                    getContext()!!, ticketPrice.isReduced,
                                    uri.getQueryParameter("number")!!,
                                    uri.getQueryParameter("message")!!
                                )
                            }
                        }

                        TicketItem.Companion.HEADER_ITEM -> {}
                    }
                }
            })
    }

    internal class TicketPrice(
        var description: CharSequence?,
        var zones: String?,
        var priceText: String?,
        var isReduced: Boolean,
        var action: String?
    ) {
        val price: CharSequence?
            get() = priceText
    }

    internal class TicketAdapter(private val mContext: Context?) :
        RecyclerView.Adapter<RecyclerView.ViewHolder?>() {
        private val mTicketItems: MutableList<TicketItem<*>> = ArrayList<TicketItem<*>>()
        private val mLayoutInflater: LayoutInflater

        init {
            this.mLayoutInflater = LayoutInflater.from(mContext)
        }

        override fun getItemViewType(position: Int): Int {
            return mTicketItems.get(position).viewType
        }

        override fun onCreateViewHolder(
            parent: ViewGroup,
            viewType: Int
        ): RecyclerView.ViewHolder {
            return when (viewType) {
                TicketItem.Companion.PRICE_ITEM -> PriceViewHolder.Companion.create(
                    mLayoutInflater,
                    parent
                )
                else -> throw IllegalArgumentException("Unknown view type: $viewType")
            }
        }

        override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
            when (getItemViewType(position)) {
                TicketItem.Companion.PRICE_ITEM -> PriceViewHolder.Companion.bindTo(
                    holder as PriceViewHolder,
                    mTicketItems.get(position).`object` as TicketPrice
                )
            }
        }

        override fun getItemCount(): Int {
            return mTicketItems.size
        }

        fun addTicket(ticket: TicketPrice?) {
            mTicketItems.add(TicketItem<TicketPrice?>(ticket, TicketItem.Companion.PRICE_ITEM))
        }

        fun getItem(position: Int): TicketItem<*> {
            return mTicketItems.get(position)
        }
    }

    class TicketItem<T>(var `object`: T?, @field:ViewType var viewType: Int) {
        @Retention(AnnotationRetention.SOURCE)
        @IntDef(HEADER_ITEM, PRICE_ITEM)
        annotation class ViewType

        companion object {
            const val HEADER_ITEM: Int = 0
            const val PRICE_ITEM: Int = 1
        }
    }

    private class PriceViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val descriptionTextView: TextView
        private val priceTextView: TextView
        private val zonesTextView: TextView

        init {
            descriptionTextView = itemView.findViewById<View?>(R.id.ticket_description) as TextView
            priceTextView = itemView.findViewById<View?>(R.id.ticket_price) as TextView
            zonesTextView = itemView.findViewById<View?>(R.id.ticket_zones) as TextView
        }

        companion object {
            fun create(inflater: LayoutInflater, parent: ViewGroup?): RecyclerView.ViewHolder {
                return PriceViewHolder(inflater.inflate(R.layout.item_ticket_price, parent, false))
            }

            fun bindTo(holder: PriceViewHolder, ticket: TicketPrice) {
                holder.descriptionTextView.setText(ticket.description)
                holder.priceTextView.setText(ticket.price)
                holder.zonesTextView.setText(ticket.zones)
            }
        }
    }

    companion object {
        const val ARG_FARE: String = "zones"

        @JvmStatic
        fun create(fare: Fare?): TicketDialogFragment {
            val b = Bundle()
            b.putParcelable(ARG_FARE, fare)
            val fragment = TicketDialogFragment()
            fragment.setArguments(b)
            return fragment
        }

        /**
         * Invokes the Messaging application.
         *
         * @param reducedPrice True if the price is reduced, false otherwise.
         * @param number       The number
         * @param message      The message
         */
        private fun sendSms(
            context: Context, reducedPrice: Boolean,
            number: String, message: String
        ) {
            val price = if (reducedPrice) "R" else "H" // Only kept for analytic purposes
            getInstance(context).event("Ticket", "Buy SMS Ticket", price)
            smsIntent(context, number, message)
        }
    }
}